src/wh_client_crypto.c

```diff
@@ -156,6 +157,284 @@ int wh_Client_RngGenerate(whClientContext* ctx, uint8_t* out, uint32_t size)
+#endif
+    if (ret == 0) {
+        if (packet->rc != 0) {
+            ret = packet->rc;
+        } else {
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+            printf("[client] out size:%d datasz:%d\n",
+                    res->sz, dataSz);
+            wh_Utils_Hexdump("[client] res_out: \n", out, res->sz);
+#endif
+            /* copy the response res_out */
+            memcpy(out, res_out, res->sz);
+            if (enc != 0) {
+                /* Update the CBC state with the last cipher text black */
+                memcpy(iv, out + last_offset, AES_IV_SIZE);
+            }
+
+        }
+    }
+    return ret;
+}
+#endif /* HAVE_AES_CBC */
+
+#ifdef HAVE_AESGCM
+int wh_Client_AesGcm(whClientContext* ctx,
+        Aes* aes, int enc,
+        uint32_t len, const uint8_t* in,
+        uint32_t iv_len, const uint8_t* iv,
+        uint32_t authin_len, const uint8_t* authin,
+        uint32_t tag_len, const uint8_t* dec_tag, uint8_t* enc_tag,
+        uint8_t* out)
+{
+    int ret = 0;
+    uint16_t group      = WH_MESSAGE_GROUP_CRYPTO;
+    uint16_t action     = WC_ALGO_TYPE_CIPHER;
+    uint16_t type       = WC_CIPHER_AES_GCM;
+    uint16_t dataSz     = 0;
+
+    uint32_t key_len    = aes->keylen;
+    const uint8_t* key  = (const uint8_t*)(aes->devKey);
+    whKeyId key_id      = WH_DEVCTX_TO_KEYID(aes->devCtx);
+
+    whPacket* packet    = (whPacket*)wh_CommClient_GetDataPtr(ctx->comm);
+
+    /* Request packet */
+    wh_Packet_cipher_aesgcm_req* req = &packet->cipherAesGcmReq;
+    uint8_t* req_in     = (uint8_t*)(req + 1);
+    uint8_t* req_key    = req_in + len;
+    uint8_t* req_iv     = req_key + key_len;
+    uint8_t* req_authin = req_iv + iv_len;
+    uint8_t* req_dec_tag= req_authin + authin_len;
+
+    uint32_t req_len    = WH_PACKET_STUB_SIZE + sizeof(*req) + len +
+                            key_len + iv_len + authin_len +
+                            ((enc == 0) ? tag_len : 0);
+
+    /* Response packet */
+    wh_Packet_cipher_aesgcm_res* res = &packet->cipherAesGcmRes;
+    uint8_t* res_out    = (uint8_t*)(res + 1);
+    uint8_t* res_enc_tag= res_out + len;
+
+    uint32_t res_len    = WH_PACKET_STUB_SIZE+ sizeof(*res) + len +
+                            ((enc == 0) ? 0: tag_len);
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+    printf("[client] AESGCM: enc:%d keylen:%d ivsz:%d insz:%d authinsz:%d authtagsz:%d reqsz:%u ressz:%u\n",
+                enc, key_len, iv_len, len, authin_len, tag_len,
+                req_len, res_len);
+    printf("[client] AESGCM: req:%p in:%p key:%p iv:%p authin:%p dec_tag:%p res:%p out:%p res_enc_tag:%p\n",
+            req, req_in, req_key, req_iv, req_authin, req_dec_tag, res, res_out, res_enc_tag);
+#endif
+    if (    (req_len > WOLFHSM_CFG_COMM_DATA_LEN) ||
+            (res_len > WOLFHSM_CFG_COMM_DATA_LEN)) {
+        /* if we're using an HSM key return BAD_FUNC_ARG */
+        if (WH_KEYID_ISERASED(key_id)) {
+            return CRYPTOCB_UNAVAILABLE;
```

Comment:
same comment as above re: considering making this an error

Reply:
yep exactly, that was what I was wondering - it is the "you got it half right" case - they set the HSM devId but not the KeyID. My original thoughts were just that if they specify the HSM devId, they are assuming operations should be RPC'ed to the HSM, full stop, and if anything were to prohibit that from happening, we should return an error instead of falling back to local crypto. IDK that is what seems right in my head. Current functionality is fine now, but we should def put a pin in this and flesh it out later.